### PR TITLE
Envoy TLS support with header imposition

### PR DESCRIPTION
**Pull request by jrajahalme**
Add API support for TLS proxying for HTTP, with support for header modifications and (k8s) secrets.

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/9486)

### CODE REVIEW

**Comment by jrajahalme** [comment]
test-me-please

**Comment by jrajahalme** [comment]
test-me-please

**Comment by jrajahalme** [comment]
test-me-please

**Comment by jrajahalme** [comment]
test-me-please

**Comment by jrajahalme** [comment]
test-me-please

**Comment by jrajahalme** [comment]
test-me-please

**Comment by jrajahalme** [comment]
test-me-please

**Comment by jrajahalme** [comment]
test-me-please

**Comment by jrajahalme** [comment]
test-me-please

**Comment by jrajahalme** [comment]
test-me-please

**Comment by jrajahalme** [comment]
test-me-please

**Comment by jrajahalme** [comment]
test-me-please

**Comment by jrajahalme** [comment]
Runtime test failed to load a container:
```Error: No such container: memcache-text-client```

**Comment by jrajahalme** [comment]
test-me-please

**Comment by jrajahalme** [comment]
test-me-please

**Comment by jrajahalme** [comment]
test-me-please

File: test/vagrant-local-start.sh [hunk 1]
```diff
@@ -0,0 +1,31 @@
+#!/bin/bash
+
+set -e
+
+export K8S_VERSION=${K8S_VERSION:-1.16}
```
Comment (aanm) [inline test/vagrant-local-start.sh]: if #9661 is merged first, this needs to be changed to `1.17`

File: pkg/defaults/defaults.go [hunk 2]
```diff
@@ -326,4 +326,8 @@ const (
 
 	// RestoreV6Addr is used as match for cilium_host v6 (router) address
 	RestoreV6Addr = "cilium.v6.internal.raw "
+
+	// CertsDirectory is the default directory used to find certificates
+	// specified in the L7 policies.
+	CertsDirectory = "/var/run/cilium/certs"
```
Comment (aanm) [inline pkg/defaults/defaults.go]: `CertsDirectory = RuntimePath + "/certs"`

File: pkg/k8s/annotate.go [hunk 3]
```diff
@@ -118,3 +121,20 @@ func (k8sCli K8sClient) AnnotateNode(nodeName string, encryptKey uint8, v4CIDR,
...
+		Resource("secrets").
+		Name(name).
+		VersionedParams(&v1.GetOptions{}, scheme.ParameterCodec).
+		Do().
+		Into(result)
+	if err != nil {
+		return nil, err
+	}
+	return result.Data, nil
+}
```
Comment (aanm) [inline pkg/k8s/annotate.go]: This also requires the following change:
```patch
diff --git a/install/kubernetes/cilium/charts/agent/templates/clusterrole.yaml b/install/kubernetes/cilium/charts/agent/templates/clusterrole.yaml
index d3b4c169f..a49ac1297 100644
--- a/install/kubernetes/cilium/charts/agent/templates/clusterrole.yaml
+++ b/install/kubernetes/cilium/charts/agent/templates/clusterrole.yaml
@@ -49,6 +49,12 @@ rules:
   - list
   - watch
   - update
+{{- if eq "k8s" .Values.global.tls.secretsBackend }}
+- apiGroups:
+  - ""
+  resources:
+  - secrets
+{{- end }}
 - apiGroups:
   - cilium.io
   resources:
diff --git a/install/kubernetes/cilium/values.yaml b/install/kubernetes/cilium/values.yaml
index 500ea8249..e17ac0f79 100644
--- a/install/kubernetes/cilium/values.yaml
+++ b/install/kubernetes/cilium/values.yaml
@@ -301,3 +301,6 @@ global:
 
   daemon:
     runPath: "/var/run/cilium"
+
+  tls:
+    secretsBackend: local
```

for which the users will be able to enable the RBAC for secrets access using

```
helm template cilium \
         --namespace=kube-system \
         --set global.tls.secretsBackend=k8s \
         ....
```

File: pkg/crypto/certificatemanager/certificate_manager.go [hunk 4]
```diff
@@ -159,3 +159,29 @@ func (m *Manager) GetTLSContext(ctx context.Context, tlsCtx *api.TLSContext) (ca
 	}
 	return "", "", "", err
 }
+
```
Comment (aanm) [inline pkg/crypto/certificatemanager/certificate_manager.go:169]: commit missing sign-off

File: pkg/crypto/certificatemanager/certificate_manager.go [hunk 5]
```diff
@@ -159,3 +159,29 @@ func (m *Manager) GetTLSContext(ctx context.Context, tlsCtx *api.TLSContext) (ca
...
+// GetSecretString returns a secret string stored in a k8s secret
```
Comment (aanm) [inline pkg/crypto/certificatemanager/certificate_manager.go:170]: This could have been split into a different PR for easier review.

File: install/kubernetes/cilium/charts/agent/templates/clusterrole.yaml [hunk 6]
```diff
@@ -18,6 +18,7 @@ rules:
   - services
   - nodes
   - endpoints
+  - secrets
```
Comment (aanm) [inline install/kubernetes/cilium/charts/agent/templates/clusterrole.yaml]: We can't blindly add this here, check my comment to make this conditional.

**Review by aanm — CHANGES_REQUESTED** [review]

File: test/vagrant-local-start.sh (hunk 1 above)
Comment (jrajahalme) [reply test/vagrant-local-start.sh]: Sure, will do when #9661 has merged :-)

File: pkg/defaults/defaults.go (hunk 2 above)
Comment (jrajahalme) [reply pkg/defaults/defaults.go]: Changed.

File: pkg/k8s/annotate.go (hunk 3 above)
Comment (jrajahalme) [reply pkg/k8s/annotate.go]: This would negate the need for `CertificatesPath`, assuming we would map a secret `<namespace>/<name>` to a path in `CertsDirectory` when `secretsBackend: local`?

Comment (jrajahalme) [reply pkg/k8s/annotate.go]: I would actually favor that!

Comment (jrajahalme) [reply pkg/k8s/annotate.go]: Done.

File: pkg/crypto/certificatemanager/certificate_manager.go (hunk 4 above)
Comment (jrajahalme) [reply pkg/crypto/certificatemanager/certificate_manager.go:169]: Added.

**Comment by jrajahalme** [comment]
test-me-please

**Comment by jrajahalme** [comment]
test-me-please

File: pkg/crypto/certificatemanager/certificate_manager.go (hunk 5 above)
Comment (jrajahalme) [reply pkg/crypto/certificatemanager/certificate_manager.go:170]: Maybe review 3 first commits separately, one-by-one?

File: install/kubernetes/cilium/charts/agent/templates/clusterrole.yaml (hunk 6 above)
Comment (jrajahalme) [reply install/kubernetes/cilium/charts/agent/templates/clusterrole.yaml]: Done.

**Comment by jrajahalme** [comment]
test-me-please

**Comment by jrajahalme** [comment]
Changed the CI k8s test to use default secret namespace to the PODs namespace.

**Comment by jrajahalme** [comment]
Super slow CI failed VM provisioning..

**Comment by jrajahalme** [comment]
test-me-please

File: test/vagrant-local-start.sh (hunk 1 above)
Comment (jrajahalme) [reply test/vagrant-local-start.sh]: Issued #9681 to fix this.

**Comment by aanm** [comment]
@jrajahalme can you rebase against master?

**Comment by jrajahalme** [comment]
test-me-please

**Comment by jrajahalme** [comment]
memcache runtime test hit an unexpected drop, likely bur to a race setting the policy. Testing again...

**Comment by jrajahalme** [comment]
test-me-please

**Comment by jrajahalme** [comment]
VM provisioning fail, testing again.

**Comment by jrajahalme** [comment]
test-me-please

File: pkg/policy/api/l4.go [hunk 7]
```diff
@@ -55,6 +55,59 @@ func (p PortProtocol) Covers(other PortProtocol) bool {
...
+	// +optional
+	Namespace string `json:"namespace,omitempty"`
+
+	// Name is the name of the secret.
+	Name string `json:"name"`
+}
+
+// Equal returns true if 'a' and 'b' have the same contents.
+func (a *Secret) Equal(b *Secret) bool {
+	return a == nil && b == nil || a != nil && b != nil && *a == *b
```
Comment (aanm) [inline pkg/policy/api/l4.go]: Why don't you have the receiver and the argument to be non pointers?

File: pkg/policy/api/l4.go [hunk 8]
```diff
@@ -55,6 +55,59 @@ func (p PortProtocol) Covers(other PortProtocol) bool {
...
+	Certificate string `json:"certificate,omitempty"`
+
+	// PrivateKey is the file name or k8s secret item name for the private key matching the certificate chain.
+	// If omitted, 'tls.key' is assumed, if it exists.
+	// If given, the item must exist.
+	PrivateKey string `json:"privateKey,omitempty"`
+}
+
+// Equal returns true if 'a' and 'b' have the same contents.
+func (a *TLSContext) Equal(b *TLSContext) bool {
```
Comment (aanm) [inline pkg/policy/api/l4.go]: Same here.

File: pkg/envoy/server.go [hunk 9]
```diff
@@ -522,34 +544,64 @@ func getHTTPRule(certManager policy.CertificateManager, h *api.PortRuleHTTP, ns
...
 	if len(headers) == 0 {
 		headers = nil
 	} else {
 		SortHeaderMatchers(headers)
 	}
-	return headers
+	if len(headerMatches) == 0 {
+		headerMatches = nil
+	} else {
+		// SortHeaderMatches(headerMatches)
```
Comment (aanm) [inline pkg/envoy/server.go:609]: it this commented by design?

File: pkg/envoy/server.go [hunk 10]
```diff
@@ -664,22 +716,27 @@ func getCiliumTLSContext(tls *policy.TLSContext) *cilium.TLSContext {
 	}
 }
 
-func GetEnvoyHTTPRules(certManager policy.CertificateManager, l7Rules *api.L7Rules, ns string) *cilium.HttpNetworkPolicyRules {
+func GetEnvoyHTTPRules(certManager policy.CertificateManager, l7Rules *api.L7Rules, ns string) (*cilium.HttpNetworkPolicyRules, bool) {
 	if len(l7Rules.HTTP) > 0 { // Just cautious. This should never be false.
+		canShortCircuit := true
```
Comment (aanm) [inline pkg/envoy/server.go:733]: please add a comment describing what does this mean

File: pkg/envoy/server.go [hunk 11]
```diff
@@ -709,16 +766,22 @@ func getPortNetworkPolicyRule(sel policy.CachedSelector, l7Parser policy.L7Parse
 		r.UpstreamTlsContext = getCiliumTLSContext(l7Rules.OriginatingTLS)
 	}
 
+	canShortCircuit := true
```
Comment (aanm) [inline pkg/envoy/server.go:785]: same here for the comment

File: pkg/envoy/server.go [hunk 12]
```diff
@@ -776,22 +839,31 @@ func getDirectionNetworkPolicy(l4Policy policy.L4PolicyMap, policyEnforced bool)
 		}
 
 		allowAll := false
+		canShortCircuit := true
```
Comment (aanm) [inline pkg/envoy/server.go:862]: same here for the comment

File: pkg/k8s/apis/cilium.io/v2/register.go [hunk 13]
```diff
@@ -1120,7 +1120,8 @@ var (
 	Secret = map[string]apiextensionsv1beta1.JSONSchemaProps{
 		"namespace": {
 			Description: "Namespace is the namespace in which the secret exists. If " +
-				"namespace is omitted, the namespace of the enclosing rule is assumed.",
+				"namespace is omitted, the namespace of the enclosing rule is assumed, " +
```
Comment (aanm) [inline pkg/k8s/apis/cilium.io/v2/register.go:1123]: Please bump `CustomResourceDefinitionSchemaVersion` in line 45

**Review by aanm — CHANGES_REQUESTED** [review]

File: pkg/policy/api/l4.go (hunk 7 above)
Comment (jrajahalme) [reply pkg/policy/api/l4.go]: This has been deleted in a later commit in the series.

File: pkg/policy/api/l4.go (hunk 8 above)
Comment (jrajahalme) [reply pkg/policy/api/l4.go]: To avoid copying the structures as would happen if they would be passed by value instead of by reference. I see that passing by value would prevent (accidental) mutations which would be nice.

File: pkg/envoy/server.go (hunk 9 above)
Comment (jrajahalme) [reply pkg/envoy/server.go:609]: yes, we would optimally sort the headerMatches to avoid unnecessary updates due to different order, but that has not been implemented yet. Will also need to consider if this sorting is needed after all, as data is in a slice, so the order only changes is the order of rules in the imported policies changes? What do you think?

File: pkg/envoy/server.go (hunk 10 above)
Comment (jrajahalme) [reply pkg/envoy/server.go:733]: Will do, how about something like this:
```
// Assume none of the rules have side-effects so that rule evaluation can
// be stopped as soon as the first allowing rule is found. 'canShortCircuit'
// is set to 'false' if any rules with side effects are encountered, causing
// all the applicable rules to be evaluated instead.
```

File: pkg/envoy/server.go (hunk 11 above)
Comment (jrajahalme) [reply pkg/envoy/server.go:785]: will do.

File: pkg/envoy/server.go (hunk 12 above)
Comment (jrajahalme) [reply pkg/envoy/server.go:862]: ditto.

File: pkg/k8s/apis/cilium.io/v2/register.go (hunk 13 above)
Comment (jrajahalme) [reply pkg/k8s/apis/cilium.io/v2/register.go:1123]: Thanks, did not know about the schema version! Will bump!

Comment (jrajahalme) [reply pkg/k8s/apis/cilium.io/v2/register.go:1123]: Done.

File: pkg/envoy/server.go (hunk 10 above)
Comment (jrajahalme) [reply pkg/envoy/server.go:733]: Added.

File: pkg/envoy/server.go (hunk 9 above)
Comment (jrajahalme) [reply pkg/envoy/server.go:609]: Added this comment:
```
		// Optimally we should sort the headerMatches to avoid
		// updating the policy if only the order of the rules
		// has changed. Right now, when 'headerMatches' is a
		// slice (rather than a map) the order only changes if
		// the order of the rules in the imported policies
		// changes, so there is minimal likelihood of
		// unnecessary policy updates.
```

File: pkg/policy/api/l4.go (hunk 8 above)
Comment (jrajahalme) [reply pkg/policy/api/l4.go]: Upon examining the call sites I recalled that the entities being compared are already pointers, so it is best to keep this `Equal()` function to avoid code repetition and keep the call sites more readable.

**Comment by jrajahalme** [comment]
test-me-please

**Comment by aanm** [comment]
@jrajahalme it's a real CI failure:
```
22:01:27  contrib/scripts/check-logging-subsys-field.sh
22:01:27  contrib/scripts/check-fmt.sh
22:01:28  Unformatted Go source code:
22:01:28  ./pkg/envoy/server.go
22:01:28  diff -u ./pkg/envoy/server.go.orig ./pkg/envoy/server.go
22:01:28  --- ./pkg/envoy/server.go.orig	2019-12-12 22:01:28.436460706 +0000
22:01:28  +++ ./pkg/envoy/server.go	2019-12-12 22:01:28.436460706 +0000
22:01:28  @@ -605,7 +605,7 @@
22:01:28   		// the order of the rules in the imported policies
22:01:28   		// changes, so there is minimal likelihood of
22:01:28   		// unnecessary policy updates.
22:01:28  -		
22:01:28  +
22:01:28   		// SortHeaderMatches(headerMatches)
22:01:28   	}
22:01:28   
22:01:28  make: *** [precheck] Error 1
22:01:28  Makefile:371: recipe for target 'precheck' failed
22:01:29  Makefile:87: recipe for target 'jenkins-precheck' failed
22:01:29  make: *** [jenkins-precheck] Error 1
```

**Comment by jrajahalme** [comment]
test-me-please

**Comment by jrajahalme** [comment]
CI errors seem unrelated, external connectivity issues (1.1.1.1), etc. Trying again.

**Comment by jrajahalme** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

File: pkg/crypto/certificatemanager/certificate_manager.go [hunk 14]
```diff
@@ -0,0 +1,196 @@
...
+			err = fmt.Errorf("Missing Secret name")
+			return "", "", "", err
+		}
+		name := tlsCtx.Secret.Name
+
+		// Give priority to local certificates
+		certPath := filepath.Join(m.rootPath, ns, name)
+		files, ioErr := ioutil.ReadDir(certPath)
+		if ioErr != nil {
+			err = fmt.Errorf("Certificates directory %s not found (%s)", certPath, ioErr)
```
Comment (joestringer) [inline pkg/crypto/certificatemanager/certificate_manager.go:91]: nit:
```suggestion
			err = fmt.Errorf("cannot read certificates directory %s  (%s)", certPath, ioErr)
```

(If it's really `not found` then the ioErr should say this. Or permission denied, or whatever the error really is.)

File: pkg/crypto/certificatemanager/certificate_manager.go [hunk 15]
```diff
@@ -0,0 +1,196 @@
...
+		privateName = tlsCtx.PrivateKey
+		privateMustExist = true
+	}
+
+	if tlsCtx.Secret != nil {
+		if tlsCtx.Secret.Namespace != "" {
+			ns = tlsCtx.Secret.Namespace
+		}
+		if tlsCtx.Secret.Name == "" {
+			err = fmt.Errorf("Missing Secret name")
```
Comment (joestringer) [inline pkg/crypto/certificatemanager/certificate_manager.go:82]: nit: Can we provide more context in this error path, either here or in the caller? Maybe my concern is overblown, but if you have multiple TLS policies and messed up one of them, it might be a little difficult to pinpoint which one you got wrong. The caller will log directionality, but not original policy UID or namespace or anything else that might help identify which policy is miswritten.

File: pkg/crypto/certificatemanager/certificate_manager.go [hunk 16]
```diff
@@ -0,0 +1,196 @@
...
+			return "", "", "", err
+		}
+		if caBytes != nil || publicBytes != nil || privateBytes != nil {
+			return ca, public, private, nil
+		}
+		if err != nil {
+			err = fmt.Errorf("certificates not found locally in %s nor in k8s secret %s/%s ", m.rootPath, ns, name)
+			return "", "", "", err
+		}
+		err = fmt.Errorf("certificates not found in k8s secret %s/%s", ns, name)
```
Comment (joestringer) [inline pkg/crypto/certificatemanager/certificate_manager.go:165]: Is there any meaningful distinction between the error on line 162 and the error here, or could we unconditionally use the error on line 162 to replace lines 161-165?

It seems to me like if either local or k8s-based certificate detection was successful (and TLS is configured in the specified policy) then this function would have already returned by this point. It's the fact that neither provided certificates that leads us to this error condition.

Also the path should probably be `certPath` rather than `m.rootPath`, as that's the full path that was searched for the local certs.

File: pkg/option/config.go [hunk 17]
```diff
@@ -89,6 +89,10 @@ const (
 	// BPFRoot is the Path to BPF filesystem
 	BPFRoot = "bpf-root"
 
+	// CertsDirectory is the root directory used to find out certificates used
+	// in L7 HTTPs policy enforcement.
+	CertsDirectory = "certificates-directory"
```
Comment (joestringer) [inline pkg/option/config.go:94]: nit: Should this CLI arg be more specific, like `policy-certificates-dir`, to ensure there's no mixups with the kvstore certs?

File: pkg/envoy/server.go [hunk 18]
```diff
@@ -456,12 +478,34 @@ func getL7Rule(l7 *api.PortRuleL7) *cilium.L7NetworkPolicyRule {
...
+			log.WithError(err).Debug("HeaderMatches: Using a default value due to k8s secret not being available")
+			err = nil
+		}
+	}
+
+	return value, err
 }
 
-func getHTTPRule(h *api.PortRuleHTTP) (headers []*envoy_api_v2_route.HeaderMatcher, ruleRef string) {
+func getHTTPRule(certManager policy.CertificateManager, h *api.PortRuleHTTP, ns string) (*cilium.HttpNetworkPolicyRule, bool) {
```
Comment (joestringer) [inline pkg/envoy/server.go:506]: Oh, were we just ignoring the rule references string before?

:+1: if it's unused.

File: pkg/policy/repository.go [hunk 19]
```diff
@@ -15,22 +15,61 @@
...
+	GetTLSContext(ctx context.Context, tls *api.TLSContext, defaultNs string) (ca, public, private string, err error)
+	GetSecretString(ctx context.Context, secret *api.Secret, defaultNs string) (string, error)
+}
+
+// PolicyContext is an interface policy resolution functions use to access the Repository.
+// This way testing code can run without mocking a full Repository.
+type PolicyContext interface {
+	GetSelectorCache() *SelectorCache
+	GetTLSContext(tls *api.TLSContext) (ca, public, private string, err error)
+	GetEnvoyHTTPRules(l7Rules *api.L7Rules) (*cilium.HttpNetworkPolicyRules, bool)
```
Comment (joestringer) [inline pkg/policy/repository.go:48]: nit: I think it'd be useful to document particularly the `canShortcut` bool here in context of what users or implementations of this interface can and should do. The whole interface could do with a bit of documentation.

File: pkg/envoy/server.go [hunk 20]
```diff
@@ -702,22 +855,35 @@ func getDirectionNetworkPolicy(l4Policy policy.L4PolicyMap, policyEnforced bool)
...
-					break
 				}
 
 				pnp.Rules = append(pnp.Rules, rule)
 			}
 		}
 
+		// Short-circuit rules if a rule allows all and all other rules can be short-circuited
+		if allowAll && canShortCircuit {
+			log.Info("Short circuiting HTTP rules due to rule allowing all and no other rules needing attention")
```
Comment (joestringer) [inline pkg/envoy/server.go:883]: This looks like a debug-level log, I don't think it's meaningful to most users and it's prone to repeating itself in the logs when a policy is evaluated.

File: pkg/k8s/apis/cilium.io/v2/register.go [hunk 21]
```diff
@@ -1128,10 +1169,69 @@ var (
...
+			Description: "Name identifies the header.",
+			Type:        "string",
+		},
+		"secret": {
+			Description: "Secret refers to a secret that contains the value that must be present in the request.",
+			Type:        "object",
+			Properties:  Secret,
+		},
+		"value": {
+			Description: "Value containst the header value that must be present in the request. If both Secret " +
```
Comment (joestringer) [inline pkg/k8s/apis/cilium.io/v2/register.go:1227]:
```suggestion
			Description: "Value contains the header value that must be present in the request. If both Secret " +
```

File: pkg/policy/api/utils.go [hunk 22]
```diff
@@ -44,6 +45,25 @@ func (h *PortRuleHTTP) Equal(o PortRuleHTTP) bool {
...
+}
+
+// Equal returns true if both HTTP rules are equal
+func (h *HeaderMatch) Equal(o *HeaderMatch) bool {
+	if h.Mismatch != o.Mismatch ||
+		h.Name != o.Name ||
+		h.Value != o.Value ||
+		h.Secret != nil && o.Secret == nil ||
+		h.Secret == nil && o.Secret != nil ||
+		h.Secret != nil && o.Secret != nil && *h.Secret != *o.Secret {
```
Comment (joestringer) [inline pkg/policy/api/utils.go:64]: Can this use `!h.Secret.Equal(o.Secret)`? (Ignore if this is autogenerated..)

File: pkg/policy/api/l4.go [hunk 23]
```diff
@@ -63,6 +102,26 @@ type PortRule struct {
 	// +optional
 	Ports []PortProtocol `json:"ports,omitempty"`
 
+	// TerminatingTLS is the TLS context for the connection terminated by
+	// the L7 proxy.  For egress policy this specifies the server-side TLS
+	// parameters to be applied on the connections originated from the local
+	// POD and terminated by the L7 proxy. For ingress policy this specifies
```
Comment (joestringer) [inline pkg/policy/api/l4.go:108]: nit: This API wording usually uses the Cilium `endpoint` terminology rather than `POD`. (Same below)

**Review by joestringer — APPROVED** [review]
LGTM!

I have a bunch of super trivial feedback below that could be addressed in a follow-up PR, the only one that would be meaningful for the next RC would be if we wish to change the CLI arg name.

**Comment by joestringer** [comment]
@jrajahalme merged :tada:

At this point it's probably more useful to merge this and get rc2 out and allow people to try it out and test it. It will be easier to iterate on in the master tree.

Please do follow up on the remaining feedback items above in a follow up PR.

File: pkg/envoy/server.go (hunk 18 above)
Comment (jrajahalme) [reply pkg/envoy/server.go:506]: Yep, was unused.

File: pkg/crypto/certificatemanager/certificate_manager.go (hunk 14 above)
Comment (jrajahalme) [reply pkg/crypto/certificatemanager/certificate_manager.go:91]: Fixed.

File: pkg/crypto/certificatemanager/certificate_manager.go (hunk 15 above)
Comment (jrajahalme) [reply pkg/crypto/certificatemanager/certificate_manager.go:82]: Added "Required" fields to the k8s schema validation code (pkg/k8s/apis/cilium.io/v2/register.go), which takes care of this when used with k8s.

File: pkg/option/config.go (hunk 17 above)
Comment (jrajahalme) [reply pkg/option/config.go:94]: Maybe, will need to discuss this more. We could also generalizer this to all "secrets", not just certificates. In that case a name like "local-secrets-dir" would be a better match?

File: pkg/policy/repository.go (hunk 19 above)
Comment (jrajahalme) [reply pkg/policy/repository.go:48]: Added documentation to all functions in this interface.

File: pkg/envoy/server.go (hunk 20 above)
Comment (jrajahalme) [reply pkg/envoy/server.go:883]: Right, changed to `Debug`

File: pkg/k8s/apis/cilium.io/v2/register.go (hunk 21 above)
Comment (jrajahalme) [reply pkg/k8s/apis/cilium.io/v2/register.go:1227]: Fixed.

File: pkg/policy/api/utils.go (hunk 22 above)
Comment (jrajahalme) [reply pkg/policy/api/utils.go:64]: `Secret.Equal()` did not exist, but I factored it out of this for clarity.

File: pkg/policy/api/l4.go (hunk 23 above)
Comment (jrajahalme) [reply pkg/policy/api/l4.go:108]: Changed, thanks!

File: pkg/crypto/certificatemanager/certificate_manager.go (hunk 16 above)
Comment (jrajahalme) [reply pkg/crypto/certificatemanager/certificate_manager.go:165]: Cleaned up in the follow-up PR #9807

**Comment by jrajahalme** [comment]
@joestringer Addressed your comments in #9807
